modules/gapi/src/streaming/onevpl/demux/async_mfp_demux_data_provider.cpp

```diff
@@ -0,0 +1,783 @@
+    }
+    source_reader = createCOMPtrGuard(source_reader_tmp);
+
+    GAPI_LOG_DEBUG(nullptr, "[" << this << "] " <<
+                            "created IMFSourceReader: " << source_reader);
+
+    if (SUCCEEDED(hr)) {
+        // Ask for the first sample.
+        request_next(hr, 0, 0);
+    } else {
+        throw DataProviderSystemErrorException(HRESULT_CODE(hr), "Cannot ReadSample");
+    }
+    GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                            "initialized");
+}
+
+MFPAsyncDemuxDataProvider::~MFPAsyncDemuxDataProvider() {
+    GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                            "deinitializing");
+
+    flush();
+
+    {
+        std::unique_lock<std::mutex> l(buffer_storage_mutex);
+        GAPI_LOG_INFO(nullptr, "Clean up async storage, count: " <<
+                               worker_key_to_buffer_mapping_storage.size());
+        for (auto& buffer : worker_key_to_buffer_mapping_storage) {
+            if (buffer.second) {
+                buffer.second->Unlock();
+            }
+        }
+        worker_key_to_buffer_mapping_storage.clear();
+    }
+
+    GAPI_LOG_INFO(nullptr, "Clean didn't used up storage, count: " <<
+                           processing_key_to_buffer_mapping_storage.size());
+    for (auto& buffer : processing_key_to_buffer_mapping_storage) {
+        if (buffer.second) {
+            buffer.second->Unlock();
+        }
+    }
+    processing_key_to_buffer_mapping_storage.clear();
+
+    // release COM object before overall MFP shutdown
+    source_reader.reset();
+    source.reset();
+
+    GAPI_DbgAssert(com_interface_reference_count.load() > 0 &&
+                   "Incorrect reference counting for MFPAsyncDemuxDataProvider");
+    MFShutdown();
+    GAPI_LOG_INFO(nullptr, "[" << this << "] " <<
+                           "deinitialized");
+}
+
+
+ULONG MFPAsyncDemuxDataProvider::AddRef() {
+    // align behavior with InterlockedIncrement
+    return com_interface_reference_count.fetch_add(1) + 1;
+}
+
+ULONG MFPAsyncDemuxDataProvider::Release() {
+    auto count = com_interface_reference_count.fetch_sub(1);
+    GAPI_Assert(count != 0 && "Incorrect reference counting for MFPAsyncDemuxDataProvider");
+    count -= 1; // align behavior with InterlockedDecrement
+    return count;
+}
+
+HRESULT MFPAsyncDemuxDataProvider::QueryInterface(REFIID riid, void** ppv)
+{
+    static const QITAB qit[] =
+    {
+        QITABENT(MFPAsyncDemuxDataProvider, IMFSourceReaderCallback),
+        { 0 },
+    };
+    return QISearch(this, qit, riid, ppv);
+}
+
+
+STDMETHODIMP
+MFPAsyncDemuxDataProvider::OnReadSample(HRESULT status, DWORD stream_index,
+                                        DWORD stream_flag, LONGLONG timestamp,
+                                        IMFSample *sample_ptr) {
+    return on_read_sample_impl(status, stream_index, stream_flag,
```

Comment:
Why did you split it into 2 functions?

Reply:
Also, they have different return types: `STDMETHODIMP` and `HRESULT`

Reply:
please find out `IntrusiveAsyncDemuxDataProvider` in tests